### ISSUE TITLE
Add app to F-Droid

### ISSUE THREAD

**Issue by papjul** [breezy-weather/breezy-weather#354]
### Describe your idea

**Edit:** Given the many cons described below, it will be available as our own F-Droid repo to begin.

Main issues I see:
1) F-Droid is slow, *really* slow to push updates. We already have users reporting issues saying "I use latest version" on GitHub while they don’t. While a IzzyOnDroid user can update from GitHub, a F-Droid user won’t be able to, as it is a different key. We are mostly doing rolling releases, so it’s really a nightmare to have F-Droid lagging behind. Example of slowness witnessed no later than this week: an app I use pushed an update on Day 1. On Day 3, they noticed a critical bug and pushed a fix release. On Day 5, F-Droid pushed the initial release, not the fixed release.
    - Issues are already time-consuming, I don’t want to handle issues reported on a version that has a different behavior, or issues already fixed. As there is probably no intention to fix this slowness on F-Droid side (basically, I’m asking for update checking every 24 hours, such as what IzzyOnDroid does), best I can do here is detect if it is F-Droid version in the app and show a warning about F-Droid version being unsupported in the settings menu, such as what Tachiyomi does. However, in that case, what becomes the point of adding it in F-Droid if we don’t support it?
2) F-Droid doesn’t allow non-public secrets, this means that no API key will be included. In other words, only the following sources will be available to use directly: MET Norway, Open-Meteo (an API key will become mandatory soon), China.
    - User can add its own API key, but let’s be honest, it’s a pain and creates a lot of frustration when the user realizes after installing the app that he has to register for a key. We don’t want these complaints to be the main thing people talk about.
    - Most weather apps only have one API, such as MET Norway, so we could say “it’s fine to only have MET Norway”, but people know it’s a fork so they will keep using Geometric Weather if they don’t have access to AccuWeather API (although I find it rather unusual to prefer using an unmaintained app that comes from F-Droid rather than a maintained app coming from GitHub just to have access to an API that is not privacy-friendly at all).
3) F-Droid picks up all tags, including pre-releases (can probably be configured, must check F-Droid docs)

Main reasons why people use F-Droid are:
- It was reviewed *initially* by organization members and scanned regularly, giving a sense of security that it does nothing nasty.
    - On this point, I was the one helping getting Geometric Weather have a `fdroid` variant. My fork goes even further in making it more privacy-friendly.
        - Current location is no longer added by default
        - AccuWeather is no longer the default source for current location, and user will always be able to choose its source easily
        - Links to privacy policy of all sources are available
        - Open-Meteo (FOSS source) was added, as well as MET Norway which have good privacy policies
- It is guaranteed to match source code (although it doesn’t mean much if no one audits it)
    - On this point, we set up a GitHub CI that builds the app at each release. So we basically do the same thing as F-Droid. F-Droid sometimes patch things on their build, but it won’t be necessary in our case, as we don’t have anything to patch to be ready for F-Droid inclusion (if Geometric Weather passed inclusion, Breezy Weather will pass it even more).
- It’s a store, it handles updates for them.
    - On this point, I recommand [using Obtainium](https://github.com/ImranR98/Obtainium), that can handle both F-Droid updates for your existing apps, and updates for your GitHub apps.
    - If you don’t want to change your current store app, you can add IzzyOnDroid repo, that only has a 24-hour delay in updates.

So if you are reading this because you only want to install this app once it will be available in F-Droid, you got the reasons why using F-Droid version is a bad idea from both developer and user point, and the reasons why you shouldn’t be afraid of using the app from GitHub directly.

So, as a user of F-Droid myself, I still want to add this app to F-Droid but now that I'm on the developer side, I understand better the issues app maintainers encounter. And honestly, I’m not sure when issues listed will be cleared.

Disclaimer: some info I put in this post might be a wrong interpretation from me or have workarounds available, in that case, please let me know!

### Sum up

| Version | F-Droid official repo | IzzyOnDroid | GitHub/Obtainium |
| --- | --- | --- | --- |
| Update delay | ❌ Unpredictable | ✅ 24-hour delay max | ✅ Instant, including optional pre-releases |
| Free-tier for sources | ❌ | ✅ | ✅ |
| Match source code | ✅ | ✅ | ✅ |
| Support | ❌ | ✅ | ✅ |
| Reach lot of users | ✅ | ❌ | ❌ |
| Compatible with F-Droid store apps | ✅ | ✅ | ❌ |

The more I look at it, the more I see no point in having app in F-Droid official repo, it only brings additional troubles, and reaching more users is not a goal to me, as long as current users are happy with it. But it’s not a closed door, if the pros can outweigh the cons.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] My idea is not related to an existing feature described in the [Homepage](https://github.com/breezy-weather/breezy-weather/blob/main/docs/HOMEPAGE.md) document.
- [X] I have updated the app to **[version 4.5.4](https://github.com/breezy-weather/breezy-weather/releases/latest)**.

**Comment by julianfairfax** [comment]
I was actually about to open an issue to track this.

I must confess, I haven't read the entire thing, but I think the app would be more accessible on F-Droid, and users of it there will just have to deal with the downsides it comes with.

I look forward to seeing the app on there!

**Comment by papjul** [comment]
> I think the app would be more accessible on F-Droid

By "accessible", if you mean "reaching more users", it’s not necessarily a goal in itself. I would even say it would give me more work with more issues reported. As with most FOSS developers working for free, at first, we are developing for ourself, and after making it public, if someone else enjoys it as well or if we can pick up ideas that people would like, that’s all for the better. But if expanding the userbase just brings unnecessary work, then that’s not something I want. Or rather, not until the organization has someone reliable, responsive and willing to take care of triage.

Btw, I know it’s clear for you, but I want to make it clear for everyone else that we are talking about F-Droid official repo. App *is* available on F-Droid through IzzyOnDroid repo, without all of these problems, except that it has a 24 hours delay which is acceptable and since it’s same signature, user can still mix IzzyOnDroid and GitHub.

> users of it there will just have to deal with the downsides it comes with

In a perfect world, yes. In reality, you know that F-Droid users also need to report their own issues. And I can’t just say “install app from GitHub”, since signature is incompatible and there is no backup/restore mechanism (there is # to track this but I don’t even know how I can make it user-input-safe). And once it’s on F-Droid official repo, there is no going back. If I just stop releasing there, people will just keep an unmaintained app forever.

Having unpredictable releases and not allowing to have secrets is something unique in app stores world as far as I know, and these are two really big issues for me, as I described in first post.

**Comment by julianfairfax** [comment]
About the signature issue, reproducible builds are an option.

**Comment by papjul** [comment]
> About the signature issue, reproducible builds are an option.

Thanks for sharing, I didn’t know.

Link to doc to read: https://f-droid.org/en/docs/Reproducible_Builds/

**Comment by EmberHeartshine** [comment]
NewPipe is on both their own repo and on F-Droid. I think they just have "don't use the F-Droid variant, try the version on our repo first" into their bug report template. NP has a lot of issues with F-Droid's slowness since the YouTube site can change more frequently (and break NP) than F-Droid updates.

**Comment by papjul** [comment]
> NewPipe is on both their own repo and on F-Droid. I think they just edited "don't use the F-Droid variant, try the version on our repo first" into their bug report template. NP has a lot of issues with F-Droid's slowness since the YouTube site can change more frequently (and break NP) than F-Droid updates.

I use NewPipe myself and it’s a great example of why it’s a problem to wait for so long, although I use official APIs so Breezy Weather doesn’t break as much. If people would be fine with a F-Droid repo dedicated to Breezy Weather, that would be much easier, but it won't change much from what IzzyOnDroid does.

> A question, though: how did Geometric Weather use AccuWeather as the default API if non-public secrets are disallowed? That may be worth investigating.

They did not use any secret, all their API keys were public, everyone can look at the code and steal keys. While it’s obviously a very bad practice, it’s also against most ToS of sources (every time I request a new key for a source, they always warn me to not commit it).

**Comment by EmberHeartshine** [comment]
> They did not use any secret, all their API keys were public, everyone can look at the code and steal keys. While it’s obviously a very bad practice, it’s also against most ToS of sources (every time I request a new key for a source, they always warn me to not commit it).

Ah, lol, that would explain it. I was almost expecting something along those lines.

**Comment by papjul** [comment]
> How exactly does this app handle secrets? There are only two options: proxying requests through a server and having the key embedded into the binary. The second option is very much not secret.

It uses GitHub secrets. F-Droid doesn't allow binary blobs as it would not be FOSS, and requests through a server is even worse IMO. Since it's open source, you can just have a look at the code to know what query to make to get the server to give you all keys. It would also add another tier in the process which some can consider uselessly invasive. Of course, there is no 100% way to protect keys from getting stolen in some way, but at least, not making them public should be a minimum.

> Also: it would be really good to see a reproducible build in main F-Droid repo, even if it requires getting an API key.

That's what you all came for here. I'm pretty sure we also have the opposite kind of user if and when it will be available to F-Droid, when their issue will be closed because of "unsupported version" will say "why do you add your app to F-Droid if you don't support it?" and I agree with them on that point.

You still have to convince me:
- it's a good idea to deliver a cheap and unsupported version to some users. Even worse: some current current users of the app might switch to the unsupported F-Droid version if and when it gets released.
- there is any reason to use F-Droid official repo over IzzyOnDroid (or a custom repo) that picks up builds generated by GitHub CI.

**Comment by julianfairfax** [comment]
> The more I look at it, the more I see no point in having app in F-Droid official repo

In that case, provide your own repository. F-Droid builds apps from the source code, so it's most trustworthy. If you won't do this, then I would prefer to be able to add a repo just for this app, to keep the number of "true F-Droid apps" the highest possible.

**Comment by papjul** [comment]
> > The more I look at it, the more I see no point in having app in F-Droid official repo
> 
> In that case, provide your own repository. F-Droid builds apps from the source code, so it's most trustworthy. If you won't do this, then I would prefer to be able to add a repo just for this app, to keep the number of my "true F-Droid apps" the highest possible.

I can do that! Especially with [your CI](https://gitlab.com/julianfairfax/fdroid-repo), it should not be hard.

**Comment by papjul** [comment]
Thanks to @julianfairfax, we are now available on F-Droid! Instructions here:
https://github.com/breezy-weather/fdroid-repo/blob/main/README.md

If you are specifically interested by the repo maintained by F-Droid themselves, please follow the following milestone or, even better, contribute to tests to help move the app to stable state!
https://github.com/breezy-weather/breezy-weather/milestone/2
